File: CSLForest/cosegment_BBox/m_BranchBound.cpp
#ifndef M_BRANCHBOUND_H
#define M_BRANCHBOUND_H
#include <ctime>
#include <iostream>
#include <cmath>
#include <queue>
#include <string>
#include <vector>
#include "m_utils.h"
#include "mex.h"
using namespace std;

template<class TreeNode>
TreeNode m_branch_bound(TreeNode rootNode, double tolFac){
   double bestUb, newUb;
   TreeNode topNode, bestNode;
   priority_queue < TreeNode, vector<TreeNode>, TreeNodeCompare<TreeNode> > myQ;
   rootNode.cmpBnds();
   bestUb = rootNode.ub;
   myQ.push(rootNode);
   bestNode = rootNode;
   int itr = 0;
#ifdef RUN_DEBUG_FROM_C
   int count =0;
   time_t curTime = time(NULL);
   time_t startTime = curTime;
   time_t pastTime = curTime;
#endif

   while (1){
	  itr++;
      if (myQ.size() == 0) break;
      topNode = myQ.top(); myQ.pop();
      if (isWithinTolFactor(bestUb, topNode.lb, tolFac)) break;
	  if (itr > 100000) break;
      TreeNode left, right;
      topNode.split(&left, &right);
      left.cmpBnds();
      right.cmpBnds();

      if (!isWithinTolFactor(bestUb, left.lb, tolFac)) {
         myQ.push(left);
         newUb = left.ub;
         if (bestUb > newUb){
            bestUb = newUb;
            bestNode = left;
         }
      }

      if (!isWithinTolFactor(bestUb, right.lb, tolFac)){
         myQ.push(right);
         newUb = right.ub;
         if (bestUb > newUb){
            bestUb = newUb;
            bestNode = right;
         }
      }

#ifdef RUN_DEBUG_FROM_C
      count++;
      if ((count % 1000) == 0){
         curTime = time(NULL);
         cout << count << " " << myQ.size() << ", bestUb: " << bestUb << ", top lb: " << topNode.lb <<
            ", left: " << left.lb << ", right: " << right.lb <<
            ", time: " << difftime(curTime, pastTime) << endl;
         pastTime = curTime;
      }
#endif
   }
#ifdef RUN_DEBUG_FROM_C
   cout << "Total iters: " << count << ", time: " << difftime(time(NULL), startTime) << endl;
#endif

   return bestNode;
   //_CrtDumpMemoryLeaks();
}
#endif

File: CSLForest/cosegment_BBox/m_branch_bound.m
function [bestNode] = m_branch_bound (rootNode, tolFac, integralIms, hist_target, binWeights, PYR_LEVELS, C, K, nHistBin)
    hist_target = floor(hist_target); %mex code used ints
    myQ = PriorityQueue(logical(0));
    rootNode = rootNode.cmpBnds(integralIms, hist_target, binWeights, PYR_LEVELS, C, K, nHistBin);
    bestUb = rootNode.ub;
    myQ.add(rootNode.lb, rootNode);
    bestNode = rootNode;
    itr = 0;

    while (1)
        itr = itr + 1;
        if (myQ.size() == 0)
            break;
        end
        [priority, topNode] = myQ.pop();

        if (isWithinTolFactor(bestUb, topNode.lb, tolFac))
            break;
        end
        if (itr > 100000)
            break;
        end

        [leftWinPair, rightWinPair] = topNode.split();

        left = TreeNode1a(leftWinPair);
        right = TreeNode1a(rightWinPair);

        left = left.cmpBnds(integralIms, hist_target, binWeights, PYR_LEVELS, C, K, nHistBin);
        right = right.cmpBnds(integralIms, hist_target, binWeights, PYR_LEVELS, C, K, nHistBin);
        
        if ~isWithinTolFactor(bestUb, left.lb, tolFac)
            myQ.add(left.lb, left);
            newUb = left.ub;
            if (bestUb > newUb)
                bestUb = newUb;
                bestNode = left;
            end
        end
        
        if ~isWithinTolFactor(bestUb, right.lb, tolFac)
            myQ.add(right.lb, right);
            newUb = right.ub;
            if (bestUb > newUb)
                bestUb = newUb;
                bestNode = right;
            end
        end
    end
end

function [result] = isWithinTolFactor(bestUb, lb, tolFactor)
    if (bestUb - lb) <= tolFactor*abs(bestUb)
        result = 1;
    else
        result = 0;
    end
end
